packages/canon/src/components/Menu/Combobox.tsx

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright 2025 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  forwardRef,
+  useState,
+  useMemo,
+  useCallback,
+  useId,
+  ChangeEvent,
+  KeyboardEvent,
+  useRef,
+  useEffect,
+} from 'react';
+import clsx from 'clsx';
+import { MenuComboboxOption, MenuComboboxProps } from './types';
+import { Icon } from '@backstage/canon';
+
+const getListboxItemId = (listboxId: string, optionValue: string): string =>
+  `${listboxId}-option-${optionValue}`;
+
+// Internal component for rendering individual items
+function ComboboxItem({
+  option,
+  optionIndex,
+  value,
+  activeOptionIndex,
+  onItemActive,
+  onItemSelect,
+  listboxId,
+}: {
+  option: MenuComboboxOption;
+  optionIndex: number;
+  value?: string[];
+  activeOptionIndex: number;
+  onItemActive: (index: number) => void;
+  onItemSelect: (value: string) => void;
+  listboxId: string;
+}) {
+  const isSelected = value?.includes(option.value) ?? false;
+  const isHighlighted = optionIndex === activeOptionIndex;
+  const itemId = getListboxItemId(listboxId, option.value);
+
+  const itemRef = useRef<HTMLDivElement>(null);
+
+  // Scroll the item into view when it becomes highlighted
+  useEffect(() => {
+    if (isHighlighted && itemRef.current) {
+      itemRef.current.scrollIntoView({ block: 'nearest' });
+    }
+  }, [isHighlighted]);
+
+  return (
+    <div
+      ref={itemRef}
+      className="canon-SubmenuComboboxItem"
+      id={itemId}
+      role="option"
+      aria-selected={isSelected}
+      data-highlighted={isHighlighted ? true : undefined}
+      data-selected={isSelected ? true : undefined}
+      data-disabled={option.disabled ? true : undefined}
+      onMouseOver={() => !option.disabled && onItemActive(optionIndex)}
+      onClick={() => !option.disabled && onItemSelect(option.value)}
+    >
+      <div className="canon-SubmenuComboboxItemCheckbox">
+        {isSelected && <Icon aria-hidden="true" name="check" size={12} />}
+      </div>
+      <div className="canon-SubmenuComboboxItemLabel">{option.label}</div>
+    </div>
+  );
+}
+
+/** @public */
+export const Combobox = forwardRef<HTMLDivElement, MenuComboboxProps>(
```

Comment:
Do you have any indication on whether a combobox is on the way from Base UI? I imagine we'll need to move over to their API at some point.

This autocomplete dropdown territory is also quite complex. It tends to end up in the need to be able to load in paginated data from the server with typeahead search and other fun bits.

Reply:
Apparently it's coming in August.